CI: No module named 'boto3'

First appeared in the [scheduled run](https://github.com/nextstrain/cli/actions/runs/19903439845) on 2025-12-03.

The error seems to affect later versions of Python, though the exact version differs across jobs.

```
/opt/hostedtoolcache/Python/3.10.19/x64/lib/python3.10/site-packages/nextstrain/cli/__init__.py:20: in <module>
    from .command     import all_commands
/opt/hostedtoolcache/Python/3.10.19/x64/lib/python3.10/site-packages/nextstrain/cli/command/__init__.py:1: in <module>
    from . import (
/opt/hostedtoolcache/Python/3.10.19/x64/lib/python3.10/site-packages/nextstrain/cli/command/run.py:30: in <module>
    from .. import runner
/opt/hostedtoolcache/Python/3.10.19/x64/lib/python3.10/site-packages/nextstrain/cli/runner/__init__.py:7: in <module>
    from . import docker, conda, singularity, ambient, aws_batch
/opt/hostedtoolcache/Python/3.10.19/x64/lib/python3.10/site-packages/nextstrain/cli/runner/aws_batch/__init__.py:96: in <module>
    from . import jobs, s3
/opt/hostedtoolcache/Python/3.10.19/x64/lib/python3.10/site-packages/nextstrain/cli/runner/aws_batch/jobs.py:10: in <module>
    from ... import aws
/opt/hostedtoolcache/Python/3.10.19/x64/lib/python3.10/site-packages/nextstrain/cli/aws/__init__.py:5: in <module>
    import boto3
E   ModuleNotFoundError: No module named 'boto3'
```

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -13,6 +13,11 @@ development source code and as such may not be routinely kept up to date.
 
 # __NEXT__
 
+## Bug fixes
+
+* Updated the s3fs dependency to avoid versions starting with 2025.12.0. This
+  maintains a dependency on `boto3` which is used for various commands.
+  ([#495](https://github.com/nextstrain/cli/pull/495))
 
 # 10.4.1 (14 October 2025)
 

--- a/doc/changes.md
+++ b/doc/changes.md
@@ -16,6 +16,12 @@ development source code and as such may not be routinely kept up to date.
 (v-next)=
 ## __NEXT__
 
+(v-next-bug-fixes)=
+### Bug fixes
+
+* Updated the s3fs dependency to avoid versions starting with 2025.12.0. This
+  maintains a dependency on `boto3` which is used for various commands.
+  ([#495](https://github.com/nextstrain/cli/pull/495))
 
 (v10-4-1)=
 ## 10.4.1 (14 October 2025)

--- a/setup.py
+++ b/setup.py
@@ -118,12 +118,19 @@ setup(
         # declarations.
         #
         # Resolve the issue by using a specially-provided package extra from
-        # s3fs (first introduced with 2021.4.0) which causes them to declare an
-        # explicit dependency on aiobotocore's specially-provided package extra
-        # on boto3 so that dependency resolver can figure it out properly.
+        # s3fs (first introduced with 2021.4.0, removed in 2025.12.0) which
+        # causes them to declare an explicit dependency on aiobotocore's
+        # specially-provided package extra on boto3 so that dependency resolver
+        # can figure it out properly.
         #
-        # See <https://github.com/dask/s3fs/issues/357> and
-        # <https://github.com/nextstrain/cli/issues/133> for more background.
+        # Note that the upper limit is not future-proof and will likely cause
+        # issues down the road. There may be a better combination to use here,
+        # but that needs extra digging.
+        #
+        # More background:
+        # <https://github.com/dask/s3fs/issues/357>
+        # <https://github.com/nextstrain/cli/issues/133>
+        # <https://github.com/fsspec/s3fs/issues/994>
         #
         # What a mess.
         #
@@ -131,7 +138,7 @@ setup(
         # https://github.com/fsspec/filesystem_spec/pull/1358 that causes the
         # error described in https://github.com/fsspec/s3fs/issues/790
         "fsspec !=2023.9.1",
-        "s3fs[boto3] >=2021.04.0, !=2023.9.1",
+        "s3fs[boto3] >=2021.04.0, !=2023.9.1, <2025.12.0",
 
         # From 2.0.0 onwards, urllib3 is better typed, but not usable (given
         # our dep tree) on 3.8 and 3.9 so we use types-urllib3 there (see

